File: apps/agents/src/open-canvas/nodes/generate-path/documents.ts
import { v4 as uuidv4 } from "uuid";
import { LangGraphRunnableConfig } from "@langchain/langgraph";
import { convertPDFToText } from "../../../lib/pdf";
import { createContextDocumentMessagesOpenAI as createContextDocumentMessages } from "../../../lib/context-docs";
import { getModelFromConfigLocal } from "../../../lib/model-config.local";
import { ContextDocument } from "@opencanvas/shared/types";
import {
  BaseMessage,
  HumanMessage,
  RemoveMessage,
} from "@langchain/core/messages";
import { OC_HIDE_FROM_UI_KEY } from "@opencanvas/shared/constants";

/**
 * Checks for context documents in a human message, and if found, converts
 * them to a human message with the proper content format.
 */
export async function convertContextDocumentToHumanMessage(
  messages: BaseMessage[]
): Promise<HumanMessage | undefined> {
  const lastMessage = messages[messages.length - 1];
  const documents = lastMessage?.additional_kwargs?.documents as
    | ContextDocument[]
    | undefined;
  if (!documents?.length) {
    return undefined;
  }

  const contextMessages = await createContextDocumentMessages(
    config,
    documents
  );
  return new HumanMessage({
    id: uuidv4(),
    content: [
      ...contextMessages.flatMap((m: BaseMessage) =>
        typeof m.content !== "string" ? m.content : []
      ),
    ],
    additional_kwargs: {
      [OC_HIDE_FROM_UI_KEY]: true,
    },
  });
}

export async function fixMisFormattedContextDocMessage(
  message: HumanMessage,
  config: LangGraphRunnableConfig
) {
  if (typeof message.content === "string") {
    return undefined;
  }

  const model = getModelFromConfigLocal();
  const modelName = model.modelName;
  const newMsgId = uuidv4();
  let changesMade = false;

  if (modelName.includes("openai")) {
    const newContentPromises = message.content.map(async (m: any) => {
      if (
        m.type === "document" &&
        m.source.type === "base64" &&
        m.source.data
      ) {
        changesMade = true;
        // Anthropic format
        return {
          type: "text",
          text: await convertPDFToText(m.source.data),
        };
      } else if (m.type === "application/pdf") {
        changesMade = true;
        // Gemini format
        return {
          type: "text",
          text: await convertPDFToText(m.data),
        };
      }
      return m;
    });
    const newContent = await Promise.all(newContentPromises);
    if (changesMade) {
      return [
        new RemoveMessage({ id: message.id || "" }),
        new HumanMessage({ ...message, id: newMsgId, content: newContent }),
      ];
    }
  } else if (modelName.includes("anthropic")) {
    const newContent = message.content.map((m) => {
      if (m.type === "application/pdf") {
        changesMade = true;
        // Gemini format
        return {
          type: "document",
          source: {
            type: "base64",
            media_type: m.type,
            data: m.data,
          },
        };
      }
      return m;
    });
    if (changesMade) {
      return [
        new RemoveMessage({ id: message.id || "" }),
        new HumanMessage({ ...message, id: newMsgId, content: newContent }),
      ];
    }
  } else if (modelName.includes("google-genai") || modelName.includes("gemini")) {
    const newContent = message.content.map((m: any) => {
      if (m.type === "document") {
        changesMade = true;
        // Anthropic format
        return {
          type: "application/pdf",
          data: m.source.data,
        };
      }
      return m;
    });
    if (changesMade) {
      return [
        new RemoveMessage({ id: message.id || "" }),
        new HumanMessage({ ...message, id: newMsgId, content: newContent }),
      ];
    }
  }

  return undefined;
}

File: agents-py/agents/src/open_canvas/nodes/generate_path/documents.py
import uuid
from typing import List, Optional, Dict, Any
from langchain_core.messages import HumanMessage
from langchain_core.runnables import RunnableConfig
from shared.src.types import ContextDocument
from agents.src.utils import (
    convert_pdf_to_text,
    create_context_document_messages,
    get_model_config
)
from shared.src.constants import OC_HIDE_FROM_UI_KEY

async def convert_context_document_to_human_message(
    messages: List[HumanMessage],
    config: RunnableConfig
) -> Optional[HumanMessage]:
    if not messages:
        return None
        
    last_message = messages[-1]
    documents = last_message.additional_kwargs.get("documents", [])
    if not documents:
        return None

    context_msgs = await create_context_document_messages(config, documents)
    content = []
    for msg in context_msgs:
        if isinstance(msg.content, list):
            content.extend(msg.content)
        elif isinstance(msg.content, dict):
            content.append(msg.content)
    
    return HumanMessage(
        id=str(uuid.uuid4()),
        content=content,
        additional_kwargs={OC_HIDE_FROM_UI_KEY: True}
    )

async def fix_misformatted_context_doc_message(
    message: HumanMessage,
    config: RunnableConfig
) -> Optional[List[HumanMessage]]:
    if isinstance(message.content, str):
        return None

    model_config = get_model_config(config)
    provider = model_config.get("model_provider")
    new_msg_id = str(uuid.uuid4())
    changes_made = False
    new_content = []

    if provider == "openai":
        for item in message.content:
            if item.get("type") == "document" and item.get("source", {}).get("type") == "base64":
                text = await convert_pdf_to_text(item["source"]["data"])
                new_content.append({"type": "text", "text": text})
                changes_made = True
            elif item.get("type") == "application/pdf":
                text = await convert_pdf_to_text(item.get("data", ""))
                new_content.append({"type": "text", "text": text})
                changes_made = True
            else:
                new_content.append(item)
                
    elif provider == "anthropic":
        for item in message.content:
            if item.get("type") == "application/pdf":
                new_content.append({
                    "type": "document",
                    "source": {
                        "type": "base64",
                        "media_type": "application/pdf",
                        "data": item.get("data")
                    }
                })
                changes_made = True
            else:
                new_content.append(item)
                
    elif provider == "google-genai":
        for item in message.content:
            if item.get("type") == "document":
                new_content.append({
                    "type": "application/pdf",
                    "data": item.get("source", {}).get("data")
                })
                changes_made = True
            else:
                new_content.append(item)

    if changes_made:
        return [
            HumanMessage(
                id=new_msg_id,
                content=new_content,
                additional_kwargs=message.additional_kwargs
            )
        ]
    return None 